File: frontend/src/types/compliance.ts
/** Types for the Compliance Center. */

export type ControlStatus = 'pass' | 'fail' | 'partial' | 'not_assessed';
export type CheckScope = 'org' | 'repo';
export type CheckStatus = 'pass' | 'fail';

export interface FrameworkScore {
  readonly name: string;
  readonly display_name: string;
  readonly score: number;
  readonly controls_passing: number;
  readonly controls_total: number;
  readonly last_generated: string | null;
}

export interface ComplianceSummary {
  readonly overall_score: number;
  readonly frameworks_tracked: number;
  readonly controls_passing: number;
  readonly controls_total: number;
  readonly critical_gaps: number;
  readonly last_assessment_date: string | null;
  readonly frameworks: readonly FrameworkScore[];
}

export interface ControlItem {
  readonly control_id: string;
  readonly title: string;
  readonly description: string;
  readonly status: ControlStatus;
  readonly evidence_summary: string;
  readonly last_checked: string | null;
  readonly category: string;
}

export interface FrameworkDetail {
  readonly name: string;
  readonly display_name: string;
  readonly score: number;
  readonly controls: readonly ControlItem[];
  readonly last_generated: string | null;
}

export interface PolicyCheckResult {
  readonly check_name: string;
  readonly display_name: string;
  readonly status: CheckStatus;
  readonly scope: CheckScope;
  readonly last_checked: string;
  readonly details: string;
}

export interface PolicyCheckResults {
  readonly checks: readonly PolicyCheckResult[];
  readonly last_run: string | null;
  readonly checks_passing: number;
  readonly checks_total: number;
}

export interface DataProcessingActivity {
  readonly activity_name: string;
  readonly purpose: string;
  readonly legal_basis: string;
  readonly data_categories: readonly string[];
  readonly retention_period: string;
  readonly status: string;
}

export interface BreachChecklistItem {
  readonly item: string;
  readonly complete: boolean;
}

export interface GDPRSummary {
  readonly data_processing_activities: readonly DataProcessingActivity[];
  readonly consent_tracking_enabled: boolean;
  readonly dsr_requests_total: number;
  readonly dsr_requests_completed: number;
  readonly dsr_requests_pending: number;
  readonly breach_notification_readiness: readonly BreachChecklistItem[];
  readonly data_retention_compliant: boolean;
  readonly erasure_requests_processed: number;
  readonly last_updated: string | null;
}

export type ComplianceTab = 'overview' | 'soc2' | 'iso27001' | 'nist_csf' | 'gdpr' | 'policy';

File: backend/app/schemas/compliance.py
"""Pydantic schemas for the Compliance Center endpoints."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Compliance Summary
# ---------------------------------------------------------------------------


class FrameworkScore(BaseModel):
    """Score and metadata for a single compliance framework."""

    name: str
    display_name: str
    score: float = Field(ge=0, le=100)
    controls_passing: int
    controls_total: int
    last_generated: str | None = None


class ComplianceSummary(BaseModel):
    """Aggregate compliance posture across all frameworks."""

    overall_score: float = Field(ge=0, le=100)
    frameworks_tracked: int
    controls_passing: int
    controls_total: int
    critical_gaps: int
    last_assessment_date: str | None = None
    frameworks: list[FrameworkScore]


# ---------------------------------------------------------------------------
# Framework Controls
# ---------------------------------------------------------------------------


class ControlItem(BaseModel):
    """A single compliance control within a framework."""

    control_id: str
    title: str
    description: str
    status: str = Field(
        pattern=r"^(pass|fail|partial|not_assessed)$",
        default="not_assessed",
    )
    evidence_summary: str = ""
    last_checked: str | None = None
    category: str = ""


class FrameworkDetail(BaseModel):
    """Detailed controls for a specific compliance framework."""

    name: str
    display_name: str
    score: float = Field(ge=0, le=100)
    controls: list[ControlItem]
    last_generated: str | None = None


# ---------------------------------------------------------------------------
# Policy Checks
# ---------------------------------------------------------------------------


class PolicyCheckResult(BaseModel):
    """Result of a single automated policy check."""

    check_name: str
    display_name: str
    status: str = Field(pattern=r"^(pass|fail)$")
    scope: str = Field(pattern=r"^(org|repo)$")
    last_checked: str
    details: str = ""


class PolicyCheckResults(BaseModel):
    """Collection of policy check results."""

    checks: list[PolicyCheckResult]
    last_run: str | None = None
    checks_passing: int
    checks_total: int


# ---------------------------------------------------------------------------
# GDPR Summary
# ---------------------------------------------------------------------------


class DataProcessingActivity(BaseModel):
    """A data processing activity record for GDPR."""

    activity_name: str
    purpose: str
    legal_basis: str
    data_categories: list[str]
    retention_period: str
    status: str = "active"


class GDPRSummary(BaseModel):
    """GDPR compliance dashboard summary."""

    data_processing_activities: list[DataProcessingActivity]
    consent_tracking_enabled: bool = False
    dsr_requests_total: int = 0
    dsr_requests_completed: int = 0
    dsr_requests_pending: int = 0
    breach_notification_readiness: list[dict[str, Any]] = Field(default_factory=list)
    data_retention_compliant: bool = False
    erasure_requests_processed: int = 0
    last_updated: str | None = None
